src/mlpack/core/tree/binary_space_tree/vantage_point_split_impl.hpp

```diff
@@ -0,0 +1,289 @@
+  size_t left = begin;
+  size_t right = begin + count - 1;
+
+  // First half-iteration of the loop is out here because the termination
+  // condition is in the middle.
+
+  while (AssignToLeftSubtree(bound, data, vantagePoint, left, mu) && (left <= right))
+    left++;
+
+  while ((!AssignToLeftSubtree(bound, data, vantagePoint, right, mu)) && (left <= right) && (right > 0))
+    right--;
+
+  while (left <= right)
+  {
+    // Swap columns.
+    data.swap_cols(left, right);
+
+    // Update the indices for what we changed.
+    size_t t = oldFromNew[left];
+    oldFromNew[left] = oldFromNew[right];
+    oldFromNew[right] = t;
+
+    // See how many points on the left are correct.  When they are correct,
+    // increase the left counter accordingly.  When we encounter one that isn't
+    // correct, stop.  We will switch it later.
+    while (AssignToLeftSubtree(bound, data, vantagePoint, left, mu) && (left <= right))
+      left++;
+
+    // Now see how many points on the right are correct.  When they are correct,
+    // decrease the right counter accordingly.  When we encounter one that isn't
+    // correct, stop.  We will switch it with the wrong point we found in the
+    // previous loop.
+    while ((!AssignToLeftSubtree(bound, data, vantagePoint, right, mu)) && (left <= right))
+      right--;
+  }
+
+  Log::Assert(left == right + 1);
+
+  return left;
+}
+
+template<typename BoundType, typename MatType>
+void VantagePointSplit<BoundType, MatType>::
+SelectVantagePoint(const BoundType& bound, const MatType& data,
+    const size_t begin, const size_t count, size_t& vantagePoint, ElemType& mu)
+{
+  arma::uvec vantagePointCandidates;
+
+  // Get no more than max(maxNumSamples, count) vantage point candidates
+  GetDistinctSamples(vantagePointCandidates, maxNumSamples, begin, count);
+
+  ElemType bestSpread = 0;
+
+  //  Evaluate eache candidate
+  for (size_t i = 0; i < vantagePointCandidates.n_rows; i++)
+  {
+    arma::uvec samples;
+
+    // Get no more than max(maxNumSamples, count) random samples
+    GetDistinctSamples(samples, maxNumSamples, begin, count);
+
+    // Calculate the second moment of the distance to the vantage point candidate
+    // using these random samples
+    const ElemType spread = GetSecondMoment(bound, data, samples,
+        vantagePointCandidates[i]);
+
+    if (spread > bestSpread)
+    {
+      bestSpread = spread;
+      vantagePoint = vantagePointCandidates[i];
+      //  Calculate the median value of the distance from the vantage point candidate
+      //  to these samples
+      GetMedian(bound, data, samples, vantagePoint, mu);
+    }    
+  }
+  assert(bestSpread > 0);
+}
+
+template<typename BoundType, typename MatType>
+void VantagePointSplit<BoundType, MatType>::
+GetDistinctSamples(arma::uvec& distinctSamples, const size_t numSamples,
+    const size_t begin, const size_t upperBound)
+{
+  if (upperBound > numSamples)
+  {
+    arma::Col<size_t> samples;
```

Comment:
I think it is better to move all temporal vectors and matrices outside (i.e. define as class members). That allows to avoid unnecessary memory allocations.